README.md
html2canvas
===========

[Homepage](https://html2canvas.hertzen.com) | [Downloads](https://gitlab.com/blcksec/html2canvas/releases) | [Questions](https://gitlab.com/blcksec/html2canvas/discussions/categories/q-a)

[![Gitter](https://badges.gitter.im/Join%20Chat.svg)](https://gitter.im/blcksec/html2canvas?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge) 
![CI](https://gitlab.com/blcksec/html2canvas/workflows/CI/badge.svg?branch=master)
[![NPM Downloads](https://img.shields.io/npm/dm/html2canvas.svg)](https://www.npmjs.org/package/html2canvas)
[![NPM Version](https://img.shields.io/npm/v/html2canvas.svg)](https://www.npmjs.org/package/html2canvas)

#### JavaScript HTML renderer ####

 The script allows you to take "screenshots" of webpages or parts of it, directly on the users browser. The screenshot is based on the DOM and as such may not be 100% accurate to the real representation as it does not make an actual screenshot, but builds the screenshot based on the information available on the page.


### How does it work? ###
The script renders the current page as a canvas image, by reading the DOM and the different styles applied to the elements.

It does **not require any rendering from the server**, as the whole image is created on the **client's browser**. However, as it is heavily dependent on the browser, this library is *not suitable* to be used in nodejs.
It doesn't magically circumvent any browser content policy restrictions either, so rendering cross-origin content will require a [proxy](https://gitlab.com/blcksec/html2canvas/wiki/Proxies) to get the content to the [same origin](http://en.wikipedia.org/wiki/Same_origin_policy).

The script is still in a **very experimental state**, so I don't recommend using it in a production environment nor start building applications with it yet, as there will be still major changes made.

### Browser compatibility ###

The library should work fine on the following browsers (with `Promise` polyfill):

* Firefox 3.5+
* Google Chrome
* Opera 12+
* IE9+
* Safari 6+

As each CSS property needs to be manually built to be supported, there are a number of properties that are not yet supported.

### Usage ###

The html2canvas library utilizes `Promise`s and expects them to be available in the global context. If you wish to
support [older browsers](http://caniuse.com/#search=promise) that do not natively support `Promise`s, please include a polyfill such as
[es6-promise](https://gitlab.com/jakearchibald/es6-promise) before including `html2canvas`.

To render an `element` with html2canvas, simply call:
` html2canvas(element[, options]);`

The function returns a [Promise](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise) containing the `<canvas>` element. Simply add a promise fulfillment handler to the promise using `then`:

    html2canvas(document.body).then(function(canvas) {
        document.body.appendChild(canvas);
    });

### Building ###

You can download ready builds [here](https://gitlab.com/blcksec/html2canvas/releases).

Clone git repository:

    $ git clone git://gitlab.com/blcksec/html2canvas.git

Install dependencies:

    $ npm install

Build browser bundle

    $ npm run build

### Examples ###

For more information and examples, please visit the [homepage](https://html2canvas.hertzen.com) or try the [test console](https://html2canvas.hertzen.com/tests/).

### Contributing ###

If you wish to contribute to the project, please send the pull requests to the develop branch. Before submitting any changes, try and test that the changes work with all the support browsers. If some CSS property isn't supported or is incomplete, please create appropriate tests for it as well before submitting any code changes.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://gitlab.com/conventional-changelog/standard-version) for commit guidelines.

## [1.4.1](https://gitlab.com/blcksec/html2canvas/compare/v1.4.0...v1.4.1) (2022-01-22)


### deps

* fix source maps (#2812) ([67c5e8d](https://gitlab.com/blcksec/html2canvas/commit/67c5e8dec4b2af9260a2b5b75b3399495fd1fee9)), closes [#2812](https://gitlab.com/blcksec/html2canvas/issues/2812)

### feat

* add support for `<video>` elements (#2788) ([181d1b1](https://gitlab.com/blcksec/html2canvas/commit/181d1b1103910d6e1b5277d5c007fc5e3006c6bf)), closes [#2788](https://gitlab.com/blcksec/html2canvas/issues/2788)

### fix

* Properties x and y of BoundingRect is undefined in old browser (#2797) ([e587a82](https://gitlab.com/blcksec/html2canvas/commit/e587a82dca01d9ada78cae34fd1bdb934e547f9b)), closes [#2797](https://gitlab.com/blcksec/html2canvas/issues/2797)
* source maps (#2787) ([46db867](https://gitlab.com/blcksec/html2canvas/commit/46db86755f064828559a4b0b37310f3ae94f5494)), closes [#2787](https://gitlab.com/blcksec/html2canvas/issues/2787)



# [1.4.0](https://gitlab.com/blcksec/html2canvas/compare/v1.3.4...v1.4.0) (2022-01-01)


### feat

* use native text segmenter where available (#2782) ([6521a48](https://gitlab.com/blcksec/html2canvas/commit/6521a487d78172f7179f7c973c1a3af40eb92009)), closes [#2782](https://gitlab.com/blcksec/html2canvas/issues/2782)

### fix

* adopted stylesheets (#2785) ([74696fa](https://gitlab.com/blcksec/html2canvas/commit/74696faf47c07b48b9c9587db0b999da1c08a8be)), closes [#2785](https://gitlab.com/blcksec/html2canvas/issues/2785)
* ios text wrapping with 0 width rect (#2786) ([0476d06](https://gitlab.com/blcksec/html2canvas/commit/0476d065158c33d2020a9f602b3043e5e2f90c75)), closes [#2786](https://gitlab.com/blcksec/html2canvas/issues/2786)
* reduce SLICE_STACK_SIZE to 50k (#2784) ([1cc853a](https://gitlab.com/blcksec/html2canvas/commit/1cc853a3186853eaca00af060f651697dc3497a9)), closes [#2784](https://gitlab.com/blcksec/html2canvas/issues/2784)



## [1.3.4](https://gitlab.com/blcksec/html2canvas/compare/v1.3.3...v1.3.4) (2021-12-29)


### ci

* add ios 15.0 testing (#2780) ([d922207](https://gitlab.com/blcksec/html2canvas/commit/d9222075daaed08884491b0563fc899ee0ced731)), closes [#2780](https://gitlab.com/blcksec/html2canvas/issues/2780)

### fix

* ios 15 font rendering crash (#2645) ([ba2b1cd](https://gitlab.com/blcksec/html2canvas/commit/ba2b1cd8e9a9d7932675d7abffce1526a609e769)), closes [#2645](https://gitlab.com/blcksec/html2canvas/issues/2645)



## [1.3.3](https://gitlab.com/blcksec/html2canvas/compare/v1.3.2...v1.3.3) (2021-11-23)


### ci

* fix macos action runners (#2757) ([ed57781](https://gitlab.com/blcksec/html2canvas/commit/ed577815949b6a565df54f2101cf6f0fb731c290)), closes [#2757](https://gitlab.com/blcksec/html2canvas/issues/2757)

### fix

* "offsets" text when font is big ([fd22a01](https://gitlab.com/blcksec/html2canvas/commit/fd22a01a3c9e39293f47caaed0c0e13d2dc8170c))
* const enums (#2651) ([eeda86b](https://gitlab.com/blcksec/html2canvas/commit/eeda86bd5e81fb4e97675fe9bee3d4d15899997f)), closes [#2651](https://gitlab.com/blcksec/html2canvas/issues/2651)



## [1.3.2](https://gitlab.com/blcksec/html2canvas/compare/v1.3.1...v1.3.2) (2021-08-15)


### docs

* add warning for webgl cloning with preserveDrawingBuffer=false (#2661) ([01ed879](https://gitlab.com/blcksec/html2canvas/commit/01ed87907ad9c7688880e2c5cb8ebc22ef73a4d8)), closes [#2661](https://gitlab.com/blcksec/html2canvas/issues/2661)
* include src files on www (#2660) ([58ff000](https://gitlab.com/blcksec/html2canvas/commit/58ff0003f77d825ac027eeec95fa80c0123eaf8f)), closes [#2660](https://gitlab.com/blcksec/html2canvas/issues/2660)

### feat

* add support for data-html2canvas-debug property for debugging (#2658) ([cd0d725](https://gitlab.com/blcksec/html2canvas/commit/cd0d7258c3a93f2989d5d9ec0244ba2763ea2d23)), closes [#2658](https://gitlab.com/blcksec/html2canvas/issues/2658)

### fix

* disable transition properties (#2659) ([f143166](https://gitlab.com/blcksec/html2canvas/commit/f1431665513e0a4636fb167a241f4a0571ba728a)), closes [#2659](https://gitlab.com/blcksec/html2canvas/issues/2659)
* overflows with absolutely positioned content (#2663) ([38c6829](https://gitlab.com/blcksec/html2canvas/commit/38c682955a9299ca7785af71d8f251df799405b0)), closes [#2663](https://gitlab.com/blcksec/html2canvas/issues/2663)



## [1.3.1](https://gitlab.com/blcksec/html2canvas/compare/v1.3.0...v1.3.1) (2021-08-14)


### fix

* multi arg transition/animation duration (#2657) ([1b55ed5](https://gitlab.com/blcksec/html2canvas/commit/1b55ed5668dcbbe1c6d8d7e94736d8f2da2d31c5)), closes [#2657](https://gitlab.com/blcksec/html2canvas/issues/2657)



# [1.3.0](https://gitlab.com/blcksec/html2canvas/compare/v1.2.2...v1.3.0) (2021-08-13)


### feat

* add rtl render support (#2653) ([6947982](https://gitlab.com/blcksec/html2canvas/commit/694798284838b16882e648914da0905818aa366c)), closes [#2653](https://gitlab.com/blcksec/html2canvas/issues/2653)
* correctly break graphemes (#2652) ([437b367](https://gitlab.com/blcksec/html2canvas/commit/437b367d3ba9dfd7f9a4c8042ac8d00208c09770)), closes [#2652](https://gitlab.com/blcksec/html2canvas/issues/2652)

### fix

* correctly handle allowTaint canvas cloning (#2649) ([c378e22](https://gitlab.com/blcksec/html2canvas/commit/c378e220694c14cb7b3b4b8650a7757f8fc23c7a)), closes [#2649](https://gitlab.com/blcksec/html2canvas/issues/2649)
* finish animation/transitions for elements (#2632) ([969638f](https://gitlab.com/blcksec/html2canvas/commit/969638fb94a0a14c64a667fa6e5689f79d9f1044)), closes [#2632](https://gitlab.com/blcksec/html2canvas/issues/2632)

### test

* add letter-spacing test for zwj emoji (#2650) ([f919204](https://gitlab.com/blcksec/html2canvas/commit/f919204efa06af219f155ca279f96124bb92862b)), closes [#2650](https://gitlab.com/blcksec/html2canvas/issues/2650)


tests/rollup.config.ts
import nodeResolve from '@rollup/plugin-node-resolve';
import commonjs from '@rollup/plugin-commonjs';
import sourceMaps from 'rollup-plugin-sourcemaps';
import typescript from '@rollup/plugin-typescript';
import json from '@rollup/plugin-json';
import {resolve} from 'path';

const pkg = require('../package.json');

const banner = `/*
 * ${pkg.title} ${pkg.version} <${pkg.homepage}>
 * Copyright (c) ${new Date().getFullYear()} ${pkg.author.name} <${pkg.author.url}>
 * Released under ${pkg.license} License
 */`;

export default {
    input: `tests/testrunner.ts`,
    output: [
        {
            file: resolve(__dirname, '../build/testrunner.js'),
            name: 'testrunner',
            format: 'iife',
            banner,
            sourcemap: true
        }
    ],
    external: [],
    watch: {
        include: 'tests/**'
    },
    plugins: [
        // Allow node_modules resolution, so you can use 'external' to control
        // which external modules to include in the bundle
        // https://gitlab.com/rollup/rollup-plugin-node-resolve#usage
        nodeResolve(),
        // Allow json resolution
        json(),
        // Compile TypeScript files
        typescript({
            tsconfig: resolve(__dirname, 'tsconfig.json')
        }),
        // Allow bundling cjs modules (unlike webpack, rollup doesn't understand cjs)
        commonjs({
            include: 'node_modules/**'
        }),

        // Resolve source maps to the original source
        sourceMaps()
    ]
};


rollup.config.ts
import resolve from '@rollup/plugin-node-resolve';
import commonjs from '@rollup/plugin-commonjs';
import sourceMaps from 'rollup-plugin-sourcemaps';
import typescript from '@rollup/plugin-typescript';
import json from '@rollup/plugin-json';

const pkg = require('./package.json');

const banner = `/*!
 * ${pkg.title} ${pkg.version} <${pkg.homepage}>
 * Copyright (c) ${(new Date()).getFullYear()} ${pkg.author.name} <${pkg.author.url}>
 * Released under ${pkg.license} License
 */`;

export default {
    input: `src/index.ts`,
    output: [
        { file: pkg.main, name: pkg.name, format: 'umd', banner, sourcemap: true },
        { file: pkg.module, format: 'esm', banner, sourcemap: true },
    ],
    external: [],
    watch: {
        include: 'src/**',
    },
    plugins: [
        // Allow node_modules resolution, so you can use 'external' to control
        // which external modules to include in the bundle
        // https://gitlab.com/rollup/rollup-plugin-node-resolve#usage
        resolve(),
        // Allow json resolution
        json(),
        // Compile TypeScript files
        typescript({ sourceMap: true, inlineSources: true }),
        // Allow bundling cjs modules (unlike webpack, rollup doesn't understand cjs)
        commonjs({
            include: 'node_modules/**'
        }),

        // Resolve source maps to the original source
        sourceMaps(),
    ],
}


www/src/pages/index.js
import React from 'react';
import Link from 'gatsby-link';
import logo from '../images/logo_icon.svg';
import Layout from '../components/layout';
import Footer from '../components/footer';
import Carbon from '../components/carbon';

const codeStyle = {
    backgroundColor: '#7cb342',
    textShadow: '0 1px 0 rgba(23, 31, 35, 0.5)',
    color: '#fff',
    padding: '8px 16px',
    borderRadius: '3px',
    width: '100%',
    whiteSpace: 'pre-wrap'
};

const linkStyle = {
    padding: '4px 8px',
    margin: '10px',
    border: '2px solid #fff',
    color: '#fff'
};

export default ({data}) => {
    return (
        <Layout>
        <div
            css={{
                background: '#558b2f',
                backgroundImage: 'linear-gradient(0deg, #558b2f, #7cb342)',
                minHeight: '100vh',
                display: 'flex',
                flexDirection: 'column'
            }}
        >
            <div
                css={{
                    display: 'flex',
                    flexGrow: 1
                }}
            >
                <div
                    css={{
                        maxWidth: 960,
                        color: '#fff',
                        margin: '0 auto',
                        textAlign: 'center',
                        alignSelf: 'center'
                    }}
                >
                    <div css={{display: 'flex', alignItems: 'center', justifyContent: 'center'}}>
                        <img src={logo} alt="html2canvas" />
                        <h1>html2canvas</h1>
                    </div>
                    <h4 css={{color: 'rgba(255, 255, 255, 0.6)', fontWeight: 300}}>
                        Screenshots with JavaScript
                    </h4>

                    <div
                        css={{
                            display: 'flex',
                            justifyContent: 'center',
                            flexDirection: 'column',
                            '@media(min-width: 1000px)': {
                                flexDirection: 'row'
                            }
                        }}
                    >
                        <div
                            css={{
                                display: 'none',
                                '@media(min-width: 1000px)': {
                                    display: 'block'
                                }
                            }}
                        >
                            <h4>HTML</h4>
                            <div
                                css={{marginRight: '5px'}}
                                dangerouslySetInnerHTML={{
                                    __html: `<div class="gatsby-highlight">
      <pre class="language-html"><code><span class="token tag"><span class="token tag"><span class="token punctuation">&lt;</span>div</span> <span class="token attr-name">id</span><span class="token attr-value"><span class="token punctuation">=</span><span class="token punctuation">"</span>capture<span class="token punctuation">"</span></span><span class="token style-attr language-css"><span class="token attr-name"> <span class="token attr-name">style</span></span><span class="token punctuation">="</span><span class="token attr-value"><span class="token property">padding</span><span class="token punctuation">:</span> 10px<span class="token punctuation">;</span> <span class="token property">background</span><span class="token punctuation">:</span> #f5da55</span><span class="token punctuation">"</span></span><span class="token punctuation">&gt;</span></span>
    <span class="token tag"><span class="token tag"><span class="token punctuation">&lt;</span>h4</span><span class="token style-attr language-css"><span class="token attr-name"> <span class="token attr-name">style</span></span><span class="token punctuation">="</span><span class="token attr-value"><span class="token property">color</span><span class="token punctuation">:</span> #000<span class="token punctuation">;</span> </span><span class="token punctuation">"</span></span><span class="token punctuation">&gt;</span></span>Hello world!<span class="token tag"><span class="token tag"><span class="token punctuation">&lt;/</span>h4</span><span class="token punctuation">&gt;</span></span>
<span class="token tag"><span class="token tag"><span class="token punctuation">&lt;/</span>div</span><span class="token punctuation">&gt;</span></span>
</code></pre>
      </div>`
                                }}
                            />
                        </div>
                        <div
                            css={{
                                display: 'none',
                                '@media(min-width: 1000px)': {
                                    display: 'block'
                                }
                            }}
                        >
                            <h4>JavaScript</h4>
                            <div
                                css={{marginLeft: '5px'}}
                                dangerouslySetInnerHTML={{
                                    __html: `<div class="gatsby-highlight">
      <pre class="language-javascript"><code><span class="token function">html2canvas</span><span class="token punctuation">(</span>document<span class="token punctuation">.</span><span class="token function">querySelector</span><span class="token punctuation">(</span><span class="token string">"#capture"</span><span class="token punctuation">)</span><span class="token punctuation">)</span><span class="token punctuation">.</span><span class="token function">then</span><span class="token punctuation">(</span>canvas <span class="token operator">=&gt;</span> <span class="token punctuation">{</span>
    document<span class="token punctuation">.</span>body<span class="token punctuation">.</span><span class="token function">appendChild</span><span class="token punctuation">(</span>canvas<span class="token punctuation">)</span>
<span class="token punctuation">}</span><span class="token punctuation">)</span><span class="token punctuation">;</span>
</code></pre>
      </div>`
                                }}
                            />
                        </div>
                    </div>

                    <div css={{margin: '20px'}}>
                        <a
                            href="/documentation"
                            css={linkStyle}
                            onClick={e => {
                                e.preventDefault();
                                document.querySelector('#tryhtml2canvas').click();
                            }}
                        >
                            Try it out
                        </a>
                        <Link to={'/documentation'} css={linkStyle}>
                            Documentation
                        </Link>
                    </div>
                    <div
                        css={{
                            display: 'flex',
                            flexDirection: 'column',
                            '@media(min-width: 1000px)': {
                                flexDirection: 'row'
                            }
                        }}
                    >
                        <div
                            css={{
                                flex: 1,
                                backgroundColor: '#558b2f',
                                padding: '10px 20px',
                                borderRadius: '10px',
                                marginBottom: '10px',
                                '@media(min-width: 1000px)': {
                                    marginRight: '5px'
                                }
                            }}
                        >
                            <Carbon />
                        </div>
                        <div
                            css={{
                                flex: 1,

                                backgroundColor: '#558b2f',
                                padding: '10px 20px',
                                borderRadius: '10px',
                                textAlign: 'left',
                                marginBottom: '10px',
                                '@media(min-width: 1000px)': {
                                    marginLeft: '5px',
                                    marginRight: '5px'
                                }
                            }}
                        >
                            <h6>Install NPM</h6>
                            <code css={codeStyle}>npm install --save html2canvas</code>
                            <h6>Install Yarn</h6>
                            <code css={codeStyle}>yarn add html2canvas</code>
                            <div css={{marginTop: '25px'}}>
                                <a href={'/dist/html2canvas.js'} css={linkStyle}>
                                    html2canvas.js
                                </a>
                                <a href={'/dist/html2canvas.min.js'} css={linkStyle}>
                                    html2canvas.min.js
                                </a>
                            </div>
                            <div css={{textAlign: 'center', marginTop: '10px'}}>
                                {Math.round(data.site.siteMetadata.packageSize / 1024)}kb gzipped
                            </div>
                        </div>
                        <div
                            css={{
                                flex: 1,
                                backgroundColor: '#558b2f',
                                padding: '10px 20px',
                                borderRadius: '10px',
                                textAlign: 'left',
                                marginBottom: '10px',
                                '@media(min-width: 1000px)': {
                                    marginLeft: '5px'
                                }
                            }}
                        >
                            <h5>Connect</h5>
                            <div css={{height: '35px'}}>
                                <iframe
                                    title="Github"
                                    src="https://ghbtns.com/github-btn.html?user=blcksec&repo=html2canvas&type=star&count=true&size=large"
                                    frameBorder="0"
                                    scrolling="0"
                                    width="160px"
                                    height="30px"
                                />
                            </div>
                            <div>
                                <a
                                    className="twitter-follow-button"
                                    href="https://twitter.com/blcksec"
                                    data-show-screen-name="false"
                                    data-size="large"
                                >
                                    Follow @blcksec
                                </a>
                            </div>
                        </div>
                    </div>
                </div>
            </div>
            <Footer />
        </div>
        </Layout>
    );
};

export const query = graphql`
    query MetadataQuery {
        site {
            siteMetadata {
                packageSize
            }
        }
    }
`;


karma.conf.js
// Karma configuration
// Generated on Sat Aug 05 2017 23:42:26 GMT+0800 (Malay Peninsula Standard Time)

const path = require('path');
const simctl = require('node-simctl');
const iosSimulator = require('appium-ios-simulator');
const listenAddress = 'localhost';
const port = 9876;

const log = require('karma/lib/logger').create('launcher:MobileSafari');

module.exports = function(config) {
    // https://gitlab.com/actions/virtual-environments/blob/master/images/macos/macos-10.15-Readme.md
    const launchers = {
        Safari_IOS_9: {
            base: 'MobileSafari',
            name: 'iPhone 5s',
            platform: 'iOS',
            sdk: '9.0'
        },
        Safari_IOS_10: {
            base: 'MobileSafari',
            name: 'iPhone 5s',
            platform: 'iOS',
            sdk: '10.0'
        },
        Safari_IOS_12: {
            base: 'MobileSafari',
            name: 'iPhone 5s',
            platform: 'iOS',
            sdk: '12.4'
        },
        Safari_IOS_13: {
            base: 'MobileSafari',
            name: 'iPhone 8',
            platform: 'iOS',
            sdk: '13.7'
        },
        Safari_IOS_14: {
            base: 'MobileSafari',
            name: 'iPhone 8',
            platform: 'iOS',
            sdk: '14.4'
        },
        Safari_IOS_15_0: {
            base: 'MobileSafari',
            name: 'iPhone 13',
            platform: 'iOS',
            sdk: '15.0'
        },
        Safari_IOS_15: {
            base: 'MobileSafari',
            name: 'iPhone 13',
            platform: 'iOS',
            sdk: '15.2'
        },
        SauceLabs_IE9: {
            base: 'SauceLabs',
            browserName: 'internet explorer',
            version: '9.0',
            platform: 'Windows 7'
        },
        SauceLabs_IE10: {
            base: 'SauceLabs',
            browserName: 'internet explorer',
            version: '10.0',
            platform: 'Windows 7'
        },
        SauceLabs_IE11: {
            base: 'SauceLabs',
            browserName: 'internet explorer',
            version: '11.0',
            platform: 'Windows 7'
        },
        SauceLabs_Edge18: {
            base: 'SauceLabs',
            browserName: 'MicrosoftEdge',
            version: '18.17763',
            platform: 'Windows 10'
        },
        SauceLabs_Android4: {
            base: 'SauceLabs',
            browserName: 'Browser',
            platform: 'Android',
            version: '4.4',
            device: 'Android Emulator',
        },
        SauceLabs_iOS10_3: {
            base: 'SauceLabs',
            browserName: 'Safari',
            platform: 'iOS',
            version: '10.3',
            device: 'iPhone 7 Plus Simulator'
        },
        SauceLabs_iOS9_3: {
            base: 'SauceLabs',
            browserName: 'Safari',
            platform: 'iOS',
            version: '9.3',
            device: 'iPhone 6 Plus Simulator'
        },
        IE_9: {
            base: 'IE',
            'x-ua-compatible': 'IE=EmulateIE9',
            flags: ['-extoff']
        },
        IE_10: {
            base: 'IE',
            'x-ua-compatible': 'IE=EmulateIE10',
            flags: ['-extoff']
        },
        IE_11: {
            base: 'IE',
            flags: ['-extoff']
        },
        Safari_Stable: {
            base: 'SafariNative'
        },
        Chrome_Stable: {
            base: 'ChromeHeadless'
        },
        Firefox_Stable: {
            base: 'Firefox'
        }
    };

    const ciLauncher = launchers[process.env.TARGET_BROWSER];

    const customLaunchers = ciLauncher ? {target_browser: ciLauncher} : {
        stable_chrome: {
            base: 'ChromeHeadless'
        },
        stable_firefox: {
            base: 'Firefox'
        }
    };

    const injectTypedArrayPolyfills = function(files) {
        files.unshift({
            pattern: path.resolve(__dirname, './node_modules/js-polyfills/typedarray.js'),
            included: true,
            served: true,
            watched: false
        });
    };

    injectTypedArrayPolyfills.$inject = ['config.files'];

    const MobileSafari = function(baseBrowserDecorator, args) {
        if(process.platform !== "darwin"){
            log.error("This launcher only works in MacOS.");
            this._process.kill();
            return;
        }
        baseBrowserDecorator(this);
        this.on('start', url => {
            simctl.getDevices(args.sdk, args.platform).then(devices => {
                const d = devices.find(d => {
                    return d.name === args.name;
                });

                if (!d) {
                    log.error(`No device found for sdk ${args.sdk} with name ${args.name}`);
                    log.info(`Available devices:`, devices);
                    this._process.kill();
                    return;
                }

                return iosSimulator.getSimulator(d.udid).then(device => {
                    return simctl.bootDevice(d.udid).then(() => device);
                }).then(device => {
                    return device.waitForBoot(60 * 5 * 1000).then(() => {
                        return device.openUrl(url);
                    });
                });
            }).catch(e => {
                console.log('err,', e);
            });
        });
    };

    MobileSafari.prototype = {
        name: 'MobileSafari',
        DEFAULT_CMD: {
            darwin: '/Applications/Xcode.app/Contents/Developer/Applications/Simulator.app/Contents/MacOS/Simulator',
        },
        ENV_CMD: null,
    };

    MobileSafari.$inject = ['baseBrowserDecorator', 'args'];

    config.set({

        // base path that will be used to resolve all patterns (eg. files, exclude)
        basePath: '',


        // frameworks to use
        // available frameworks: https://npmjs.org/browse/keyword/karma-adapter
        frameworks: ['mocha', 'inline-mocha-fix'],

        // list of files / patterns to load in the browser
        files: [
            'build/testrunner.js',
            { pattern: './tests/**/*', 'watched': true, 'included': false, 'served': true},
            { pattern: './dist/**/*', 'watched': true, 'included': false, 'served': true},
            { pattern: './node_modules/**/*', 'watched': true, 'included': false, 'served': true},
        ],

        plugins: [
            'karma-*',
            {
                'framework:inline-mocha-fix': ['factory', injectTypedArrayPolyfills]
            },
            {
                'launcher:MobileSafari': ['type', MobileSafari]
            }
        ],

        // list of files to exclude
        exclude: [
        ],


        // preprocess matching files before serving them to the browser
        // available preprocessors: https://npmjs.org/browse/keyword/karma-preprocessor
        preprocessors: {
        },


        // test results reporter to use
        // possible values: 'dots', 'progress'
        // available reporters: https://npmjs.org/browse/keyword/karma-reporter
        reporters: ['dots', 'junit'],

        junitReporter: {
            outputDir: 'tmp/junit/'
        },

        // web server listen address,
        listenAddress,

        // web server port
        port,


        // enable / disable colors in the output (reporters and logs)
        colors: true,


        // level of logging
        // possible values: config.LOG_DISABLE || config.LOG_ERROR || config.LOG_WARN || config.LOG_INFO || config.LOG_DEBUG
        logLevel: config.LOG_INFO,


        // enable / disable watching file and executing tests whenever any file changes
        autoWatch: true,


        // start these browsers
        // available browser launchers: https://npmjs.org/browse/keyword/karma-launcher
        browsers: Object.keys(customLaunchers),


        customLaunchers,

        // Continuous Integration mode
        // if true, Karma captures browsers, runs the tests and exits
        singleRun: true,

        // Concurrency level
        // how many browser should be started simultaneous
        concurrency: 5,

        proxies: {
            '/dist': `http://localhost:${port}/base/dist`,
            '/node_modules': `http://localhost:${port}/base/node_modules`,
            '/tests': `http://localhost:${port}/base/tests`,
            '/assets': `http://localhost:${port}/base/tests/assets`
        },

        client: {
            mocha: {
                // change Karma's debug.html to the mocha web reporter
                reporter: 'html'
            }
        },

        captureTimeout: 300000,

        browserDisconnectTimeout: 60000,

        browserNoActivityTimeout: 1200000
    })
};
